### PR TITLE
vscode-extensions.ms-vscode-remote.remote-ssh: init at 0.48.0

### LINKED ISSUE #103: Failed to connect to the remote extension host server due to unsupported, out of date libraries like glibc (e.g. older CentOS version)
- VSCode Version: Version 1.34.0-insider (1.34.0-insider)
- Local OS Version: macOS 10.14.4 (18E226)
- Remote OS Version: Amazon Linux Bare Metal release 2012.03
- Remote Extension/Connection Type: SSH

Steps to Reproduce:

1. Cmd-Shift-P, and choose *Remote-SSH: Connect current window to host...*
2. Connection is successful. Remote tools are installed and server is running on remote
3. VS Code shows an error at right-bottom: `Failed to connect to the remote extension host server (Error: Connection error: Unauthorized client refused.)`

`AllowTcpForwarding` is enabled on the remote server.

Here is the full output from "Remote - SSH" channel:

```
SSH Resolver called for "ssh-remote+dev-server"
SSH Resolver called for host: dev-server
Setting up SSH remote "dev-server"
Using commit id "473af338e1bd9ad4d9853933da1cd9d5d9e07dc9" and quality "insider" for server
Install and start server if needed
> bash: no job control in this shell
> Found existing installation...
> Warning: bad syntax, perhaps a bogus '-'? See /usr/share/doc/procps-3.2.8/FAQ
> Found running server...
>  
> *
> * Reminder: You may only use this software with Visual Studio family products,
> * as described in the license (https://go.microsoft.com/fwlink/?linkid=2077057)
> *
>  
> bb80525e-2122-4fa5-8e4a-9d91821eeb83==45539==
"install" terminal command done
Received install output: bb80525e-2122-4fa5-8e4a-9d91821eeb83==45539==
Server is listening on port 45539
Spawning tunnel with: ssh  dev-server -N -L localhost:32546:localhost:45539
Spawned SSH tunnel between local port 32546 and remote port 45539
Waiting for ssh tunnel to be ready
Tunneling remote port 45539 to local port 32546
Resolving "ssh-remote+dev-server" to "localhost:32546", attempt: 1
```

Logs from "Log (Window)" channel:
```
[2019-05-03 18:02:25.843] [renderer1] [error] Connection error: Unauthorized client refused.: Error: Connection error: Unauthorized client refused.
    at v (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1544:9)
    at t.onControlMessage.o (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1542:930)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at a (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:48)
    at e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:94)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at S._receiveMessage (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:330:65)
    at S._socketDisposables.push._socketReader.onMessage.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:327:243)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at g.acceptChunk (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:801)
    at _register._socket.onData.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:153)
    at Socket.t (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:332:235)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at TCP.onread (net.js:636:20)
[2019-05-03 18:02:25.844] [renderer1] [error] Cannot read property 'extensions' of null: TypeError: Cannot read property 'extensions' of null
    at e.<anonymous> (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:3850:751)
    at Generator.next (<anonymous>)
    at r (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:34:454)
    at process._tickCallback (internal/process/next_tick.js:68:7)
[2019-05-03 18:02:25.844] [renderer1] [error] Connection error: Unauthorized client refused.: Error: Connection error: Unauthorized client refused.
    at v (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1544:9)
    at t.onControlMessage.o (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:1542:930)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at a (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:48)
    at e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:321:94)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at S._receiveMessage (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:330:65)
    at S._socketDisposables.push._socketReader.onMessage.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:327:243)
    at d.fire (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:77:818)
    at g.acceptChunk (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:801)
    at _register._socket.onData.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:323:153)
    at Socket.t (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:332:235)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at TCP.onread (net.js:636:20)
[2019-05-03 18:02:28.207] [renderer1] [error] Could not fetch environment: Error: Could not fetch environment
    at d.getEnvironment.then.e (file:///Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.main.js:3813:413)
```

Does this issue occur when you try this locally?: Yes
Does this issue occur when you try this locally and all extensions are disabled?: Yes

### LINKED ISSUE #939: feature-request: vs-code server
VScode integration is awesome and super useful. I'd love to be able to configure extensions when in remote ssh server mode, too. Right now [this workaround is good](https://github.com/microsoft/vscode-remote-release/issues/648#issuecomment-503148523) for getting it working in general,  using `nodejs-12_x` instead due to updates, but still have to manually install each extension & then restart. Not very hard, but would be super slick to customize the remote, too! Let me know if I should put this in `nixpkgs` instead, although figured it had to live in userspace.

### PR DESCRIPTION
**Pull request by tbenst**
###### Motivation for this change
Currently, using VSCode extension remote-ssh with NixOS host requires an [annoying patch](https://github.com/microsoft/vscode-remote-release/issues/648#issuecomment-503148523). This client-side patch changes the package so the host node (as found by `which node`)  is *always* used. 

Main limitation: prior to patch, you could remote-ssh to Ubuntu without `nodejs_12` installed, and it would work by locally installing node binary. With this patch, you'll need to first `sudo apt install nodejs` or install nodejs in userspace via another means.

Benefit: as long as a NixOS box has `nodejs-12_x` installed, it now works out-of-the-box with `remote-ssh`. As does any other old OS, so this effectively fixes https://github.com/microsoft/vscode-remote-release/issues/103.

Closes https://github.com/rycee/home-manager/issues/939

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers
Not maintainers, but following people may have input:
cc @simonchatts @luxferresum @aanderse @Mic92

### CODE REVIEW

File: pkgs/misc/vscode-extensions/remote-ssh/default.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+let
+  inherit (vscode-utils) buildVscodeMarketplaceExtension;
+  
+  # patch runs on remote machine hence use of which
+  patch = ''
+    f="/home/''$USER/.vscode-server/bin/''$COMMIT_ID/node"
+    echo PATCH: replacing ''$f with ''$(which node)
+    rm ''$f
+    ln -s ''$(which node) ''$f
```
Comment (Mic92) [inline pkgs/misc/vscode-extensions/remote-ssh/default.nix]: To solve the ubuntu problem, could this not only symlink node, if it is actually found on the remote host?

**Comment by Mic92** [comment]
Also mentioned in the wiki: https://nixos.wiki/wiki/Vscode

**Comment by tbenst** [comment]
Hey @Mic92, apologies but this pull request has a mistake--I neglected to escape the `${` so trigger template javascript. My latest https://github.com/tbenst/nixpkgs/tree/remote-ssh has the fix (including one for useLocalExtensions as well) and is fully functional. Sorry for the hassle

**Comment by Mic92** [comment]
Can you make a pull request with the fix?

**Pull request merged**